Add implementation and interface source to the reference driver

panglesd opened the pull request:
This PR to review before adding it to the PR on the official odoc repo.

panglesd commented:
@julow

Julow commented on doc/driver.md:
```diff
@@ -224,6 +235,11 @@ let odoc_libraries = [
 
+let source_dir_of_odoc_lib lib =
+  match String.split_on_char '_' lib with
+  | "odoc" :: s -> Some Fpath.(v ".." / "src" / String.concat "_" s)
```
That's not the ideal approach because it won't work on external dependencies.

It would be great if the reference driver could use `ocamlfind` for Odoc itself too. This would allow to write a general way to find the source files. (which `source_files_of_odoc_module` is already doing)

Calling `ocamlfind query` on Odoc's libraries should return something in the `_build/install` directory. That's the better things to build the doc from.

panglesd replied:
I agree that it would be better, however all my attempts at using `ocamlfind` (which I don't know well) for `odoc` resulted in `Not found` answers. I can try to investigate a bit more.

I think that it is good to keep the driver simple, and not that bad to show that implementation and interface are optionals.

Julow replied:
`ocamlfind` is working as expected from within the driver:

```
$ ocamlfind query odoc
/home/juloo/w/odoc/_build/install/default/lib/odoc
```

This would allow to treat odoc as any other package and to implement a general function for finding the sources. I think such a function would be more interesting to the reader.

It wouldn't make the driver more complex but it's definitely a lot of work (basically rewrite the entire driver).
Let's not do this.

Julow reviewed (commented):
I get diffs saying that many modules are no longer built. Did you do something special before running the driver ?

panglesd commented:
> I get diffs saying that many modules are no longer built. Did you do something special before running the driver ?

No, nothing special...
I have some diffs, but I don't think they mean that modules are no longer built:

```diff
File "doc/driver.md", line 1, characters 0-0:
diff --git a/_build/default/doc/driver.md b/_build/default/doc/driver.md.corrected
index 1e19ea2a5..e823d595e 100644
--- a/_build/default/doc/driver.md
+++ b/_build/default/doc/driver.md.corrected
@@ -505,157 +505,9 @@ Let's see if there was any output from the `odoc` invocations:
 - : string list = [""]
 # !link_output;;
 - : string list =
-[""; "'../src/odoc/bin/main.exe' 'link' 'odoc_xref2.odoc' '-I' '.'";
- "odoc_xref2.odoc: File \"set.mli\", line 208, characters 16-32:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_elt_opt Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"set.mli\", line 204, characters 16-28:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_elt Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"set.mli\", line 189, characters 16-23:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(Make) Couldn't find \"Make\"";
- "odoc_xref2.odoc: File \"map.mli\", line 331, characters 16-24:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).map Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 248, characters 16-36:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding_opt Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 242, characters 16-32:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 223, characters 16-23:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(Make) Couldn't find \"Make\"";
- "odoc_xref2.odoc: File \"map.mli\", line 331, characters 16-24:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).map Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 248, characters 16-36:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding_opt Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 242, characters 16-32:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 223, characters 16-23:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(Make) Couldn't find \"Make\"";
- "odoc_xref2.odoc: File \"map.mli\", line 331, characters 16-24:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).map Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 248, characters 16-36:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding_opt Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 242, characters 16-32:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 223, characters 16-23:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(Make) Couldn't find \"Make\"";
- "odoc_xref2.odoc: File \"map.mli\", line 331, characters 16-24:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).map Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 248, characters 16-36:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding_opt Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 242, characters 16-32:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 223, characters 16-23:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(Make) Couldn't find \"Make\"";
- "odoc_xref2.odoc: File \"map.mli\", line 331, characters 16-24:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).map Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 248, characters 16-36:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding_opt Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 242, characters 16-32:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 223, characters 16-23:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(Make) Couldn't find \"Make\"";
- "odoc_xref2.odoc: File \"map.mli\", line 331, characters 16-24:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).map Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 248, characters 16-36:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding_opt Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 242, characters 16-32:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 223, characters 16-23:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(Make) Couldn't find \"Make\"";
- "odoc_xref2.odoc: File \"map.mli\", line 331, characters 16-24:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).map Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 248, characters 16-36:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding_opt Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 242, characters 16-32:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding Couldn't find \"S\"";
- "odoc_xref2.odoc: File \"map.mli\", line 223, characters 16-23:";
- "odoc_xref2.odoc: Warning: Failed to resolve reference unresolvedroot(Make) Couldn't find \"Make\"";
- "'../src/odoc/bin/main.exe' 'link' 'odoc_model.odoc' '-I' '.'";
- "odoc_model.odoc: File \"map.mli\", line 331, characters 16-24:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).map Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 248, characters 16-36:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding_opt Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 242, characters 16-32:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 223, characters 16-23:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(Make) Couldn't find \"Make\"";
- "odoc_model.odoc: File \"map.mli\", line 331, characters 16-24:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).map Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 248, characters 16-36:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding_opt Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 242, characters 16-32:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 223, characters 16-23:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(Make) Couldn't find \"Make\"";
- "odoc_model.odoc: File \"map.mli\", line 331, characters 16-24:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).map Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 248, characters 16-36:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding_opt Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 242, characters 16-32:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 223, characters 16-23:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(Make) Couldn't find \"Make\"";
- "odoc_model.odoc: File \"map.mli\", line 331, characters 16-24:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).map Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 248, characters 16-36:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding_opt Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 242, characters 16-32:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 223, characters 16-23:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(Make) Couldn't find \"Make\"";
- "odoc_model.odoc: File \"map.mli\", line 331, characters 16-24:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).map Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 248, characters 16-36:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding_opt Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 242, characters 16-32:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 223, characters 16-23:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(Make) Couldn't find \"Make\"";
- "odoc_model.odoc: File \"map.mli\", line 331, characters 16-24:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).map Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 248, characters 16-36:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding_opt Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 242, characters 16-32:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 223, characters 16-23:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(Make) Couldn't find \"Make\"";
- "odoc_model.odoc: File \"map.mli\", line 331, characters 16-24:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).map Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 248, characters 16-36:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding_opt Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 242, characters 16-32:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 223, characters 16-23:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(Make) Couldn't find \"Make\"";
- "odoc_model.odoc: File \"map.mli\", line 331, characters 16-24:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).map Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 248, characters 16-36:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding_opt Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 242, characters 16-32:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 223, characters 16-23:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(Make) Couldn't find \"Make\"";
- "odoc_model.odoc: File \"map.mli\", line 331, characters 16-24:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).map Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 248, characters 16-36:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding_opt Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 242, characters 16-32:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 223, characters 16-23:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(Make) Couldn't find \"Make\"";
- "odoc_model.odoc: File \"map.mli\", line 331, characters 16-24:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).map Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 248, characters 16-36:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding_opt Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 242, characters 16-32:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_binding Couldn't find \"S\"";
- "odoc_model.odoc: File \"map.mli\", line 223, characters 16-23:";
- "odoc_model.odoc: Warning: Failed to resolve reference unresolvedroot(Make) Couldn't find \"Make\"";
- "'../src/odoc/bin/main.exe' 'link' 'odoc_examples.odoc' '-I' '.'";
- "odoc_examples.odoc: File \"set.mli\", line 208, characters 16-32:";
- "odoc_examples.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_elt_opt Couldn't find \"S\"";
- "odoc_examples.odoc: File \"set.mli\", line 204, characters 16-28:";
- "odoc_examples.odoc: Warning: Failed to resolve reference unresolvedroot(S).min_elt Couldn't find \"S\"";
- "odoc_examples.odoc: File \"set.mli\", line 189, characters 16-23:";
- "odoc_examples.odoc: Warning: Failed to resolve reference unresolvedroot(Make) Couldn't find \"Make\"";
+[""; "'../src/odoc/bin/main.exe' 'link' 'odoc_odoc.odoc' '-I' '.'";
+ "odoc_odoc.odoc: File \"src/odoc/resolver.mli\", line 30, characters 4-43:";
+ "odoc_odoc.odoc: Warning: Failed to resolve reference unresolvedroot(Fs).Directory.t Couldn't find \"Fs\"";
  "'../src/odoc/bin/main.exe' 'link' 'page-deps.odoc' '-I' '.'";
  "page-deps.odoc: File \"src/fmt.mli\", line 6, characters 4-13:";
  "page-deps.odoc: Warning: Failed to resolve reference unresolvedroot(Format) Couldn't find \"Format\"";
@@ -664,14 +516,24 @@ Let's see if there was any output from the `odoc` invocations:
  "page-deps.odoc: File \"src/fpath.mli\", line 7, characters 59-71:";
  "page-deps.odoc: Warning: Failed to resolve reference unresolvedroot(Set) Couldn't find \"Set\"";
  "page-deps.odoc: File \"src/fpath.mli\", line 7, characters 28-52:";
- "page-deps.odoc: Warning: Failed to resolve reference unresolvedroot(file_exts) Couldn't find \"file_exts\""]
+ "page-deps.odoc: Warning: Failed to resolve reference unresolvedroot(file_exts) Couldn't find \"file_exts\"";
+ "'../src/odoc/bin/main.exe' 'link' 'page-stdlib.odoc' '-I' '.'";
+ "page-stdlib.odoc: File \"library_mlds/stdlib.mld\", line 199, characters 0-29:";
+ "page-stdlib.odoc: Warning: Failed to resolve reference unresolvedroot(Reg_with_debug_info) Parent_module: Lookup failure (root module): Reg_with_debug_info";
+ "page-stdlib.odoc: File \"library_mlds/stdlib.mld\", line 198, characters 0-30:";
+ "page-stdlib.odoc: Warning: Failed to resolve reference unresolvedroot(Reg_availability_set) Parent_module: Lookup failure (root module): Reg_availability_set";
+ "page-stdlib.odoc: File \"library_mlds/stdlib.mld\", line 76, characters 0-29:";
+ "page-stdlib.odoc: Warning: Failed to resolve reference unresolvedroot(Compute_ranges_intf) Parent_module: Lookup failure (root module): Compute_ranges_intf";
+ "page-stdlib.odoc: File \"library_mlds/stdlib.mld\", line 75, characters 0-24:";
+ "page-stdlib.odoc: Warning: Failed to resolve reference unresolvedroot(Compute_ranges) Parent_module: Lookup failure (root module): Compute_ranges";
+ "page-stdlib.odoc: File \"library_mlds/stdlib.mld\", line 21, characters 0-24:";
+ "page-stdlib.odoc: Warning: Failed to resolve reference unresolvedroot(Available_regs) Parent_module: Lookup failure (root module): Available_regs";
+ "'../src/odoc/bin/main.exe' 'link' 'page-odoc_for_authors.odoc' '-I' '.'";
+ "page-odoc_for_authors.odoc: File \"odoc_for_authors.mld\", line 472, character 59 to line 473, character 18:";
+ "page-odoc_for_authors.odoc: Warning: Failed to resolve reference unresolvedroot(Features).canonical Couldn't find page \"Features\""]
 # !generate_output;;
 - : string list =
 ["";
- "'../src/odoc/bin/main.exe' 'html-generate' 'odoc_xref_test.odocl' '-o' 'html' '--theme-uri' 'odoc' '--support-uri' 'odoc'";
- "odoc_xref_test.odocl: Warning, resolved hidden path: Odoc_model__Lang.Signature.t";
- "'../src/odoc/bin/main.exe' 'html-generate' 'odoc_document.odocl' '-o' 'html' '--theme-uri' 'odoc' '--support-uri' 'odoc'";
- "odoc_document.odocl: Warning, resolved hidden path: Odoc_document__Types.Inline.one";
  "'../src/odoc/bin/main.exe' 'html-generate' 'odoc_examples.odocl' '-o' 'html' '--theme-uri' 'odoc' '--support-uri' 'odoc'";
  "odoc_examples.odocl: Warning, resolved hidden path: Odoc_examples__.Unexposed.t"]
 ```

I realized that I forgot to update the `driver.mld`. Is it meant to be updated by hand or is there a dedicated automatic tool in odoc?

Julow commented:
I see the same diffs on master so it's probably fine but it means we can't use that output to make sense of whether your patch is doing anything. 

I would remove the whole logs thing and just do:

    ```sh
    $ find html | sort
    ```

panglesd commented:
> ` $ find html | sort `

That's more than 3000 lines of output. I will try to have something as meaningful but smaller.

(A way to test that my patch is doing something is to open the generated pages at `_build/default/doc/html/odoc/index.html`, but that is not very automatic...)

Julow reviewed (commented):
I get a lot of diffs when I run it.

Be careful if you merge this on the current target `render-source-file-2`, the upstream branch has been force pushed since.
